client/components/close-on-escape/index.jsx

```diff
@@ -0,0 +1,64 @@
+/**
+ * External dependencies
+ */
+import { noop, last, filter } from 'lodash';
+import { Component, PropTypes } from 'react';
+
+let components = [];
+
+function onKeydown( event ) {
+	if ( components.length > 0 && event.keyCode === 27 && ! isInput( event.target ) ) { // ESC
+		const component = last( components );
+		component.onEscape();
+	}
+}
+
+function isInput( element ) {
+	return -1 !== [ 'INPUT', 'TEXTAREA' ].indexOf( element.nodeName );
+}
+
+function addKeydownListener() {
+	document.addEventListener( 'keydown', onKeydown, true );
+}
+
+function removeKeydownListener() {
+	document.removeEventListener( 'keydown', onKeydown, true );
+}
+
+function startCloseOnEscForComponent( component, onEscape ) {
+	components.push( { component, onEscape } );
+	if ( components.length > 0 ) {
+		addKeydownListener();
+	}
+}
+
+function stopCloseOnEscForComponent( component ) {
+	components = filter( components, item => item.component !== component );
+	if ( components.length < 1 ) {
+		removeKeydownListener();
+	}
+}
+
+export default class CloseOnEscape extends Component {
```

Comment:
Oh yeah! it was originally but it turned out that a child component would be more suitable (props to @dmsnell on that suggestion!)
I just need to update the PR description to reflect those changes - since I rebased over the original commits!